Placeholder for PodHasNetwork condition (#35704)
kubernetes/website: content/en/docs/concepts/workloads/pods/pod-lifecycle.md

A Pod has a PodStatus, which has an array of
[PodConditions](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#podcondition-v1-core)
through which the Pod has or has not passed. Kubelet manages the following
PodConditions:

* `PodScheduled`: the Pod has been scheduled to a node.

> * `PodHasNetwork`: the Pod sandbox has been successfully created and networking configured.

Author: Is there a way to indicate within the bullet list that `PodHasNetwork` is in Alpha? We mention this in the dedicated section but when I tried to inline the `{{< feature-state for_k8s_version="v1.25" state="alpha" >}}` bits in the bullet list, the list looked quite messed up and hard to read.

Reviewer: ```suggestion
* `PodHasNetwork`: (alpha feature; must be [enabled explicitly](#pod-has-network)) the
  Pod sandbox has been successfully created and networking configured.
```

When a Pod's containers are Ready but at least one custom condition is missing or
`False`, the kubelet sets the Pod's [condition](#pod-conditions) to `ContainersReady`.

### Pod network readiness {#pod-has-network}

{{< feature-state for_k8s_version="v1.25" state="alpha" >}}

> 

Reviewer: Do folks need to enable a feature gate for this to work? (yes)

Please mention `PodHasNetworkCondition` and link to https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/

To find existing examples of use, search the English docs for `state="alpha"`.

Author: done

After a Pod gets scheduled on a node, it needs to be admitted by the Kubelet and
have any volumes mounted. Once these phases are complete, the Kubelet works with

> a runtime plugin (using [CRI](/docs/concepts/architecture/cri/)) to set up a

Reviewer: (nit)
```suggestion
a container runtime (using {{< glossary_tooltip term_id="cri" >}}) to set up a
```

runtime sandbox and configure networking for the Pod.

The `PodHasNetwork` condition is set to `False` by the Kubelet when it detects a
Pod does not have a runtime sandbox with networking configured. This occurs in
the following scenarios:
* Early in the lifecycle of the Pod when the Kubelet has not started setting up
a sandbox for the Pod using the runtime plugin.

> * Later in the lifecycle of the Pod when the Pod sandbox has been destroyed due
> to the host rebooting (without the Pod getting evicted) or the runtime sandbox
> environment rebooting (in case of MicroVM scenarios like Kata) requiring a new
> runtime sandbox to be created and networking configured again.

Reviewer: ```suggestion

* Early in the lifecycle of the Pod, when the kubelet has not yet begun to set up
  a sandbox for the Pod using the container runtime.
* Later in the lifecycle of the Pod, when the Pod sandbox has been destroyed due
  to either:
  * the node rebooting, without the Pod getting evicted
  * for container runtimes that use virtual machines for isolation, the Pod sandbox
    virtual machine rebooting, which then requires creating a new sandbox and fresh
    container network configuration.
```

When a Pod's containers are Ready but at least one custom condition is missing or
`False`, the kubelet sets the Pod's [condition](#pod-conditions) to `ContainersReady`.

### Pod network readiness {#pod-has-network}

{{< feature-state for_k8s_version="v1.25" state="alpha" >}}

After a Pod gets scheduled on a node, it needs to be admitted by the Kubelet and
have any volumes mounted. Once these phases are complete, the Kubelet works with
a container runtime (using {{< glossary_tooltip term_id="cri" >}}) to set up a
runtime sandbox and configure networking for the Pod.

The `PodHasNetwork` condition is set to `False` by the Kubelet when it detects a
Pod does not have a runtime sandbox with networking configured. This occurs in
the following scenarios:

> * Early in the lifecycle of the Pod when the Kubelet has not started setting up
> a sandbox for the Pod using the runtime plugin.

Reviewer 2: ```suggestion

* Early in the lifecycle of the Pod when the kubelet has not started setting up
  a sandbox for the Pod using the runtime plugin.
```

* Early in the lifecycle of the Pod, when the kubelet has not yet begun to set up
  a sandbox for the Pod using the container runtime.
* Later in the lifecycle of the Pod, when the Pod sandbox has been destroyed due
  to either:
  * the node rebooting, without the Pod getting evicted
  * for container runtimes that use virtual machines for isolation, the Pod sandbox
    virtual machine rebooting, which then requires creating a new sandbox and fresh
    container network configuration.

> The `PodHasNetwork` condition is set to `True` by the Kubelet after the

Reviewer 2: ```suggestion
The `PodHasNetwork` condition is set to `True` by the kubelet after the
```

successful completion of sandbox creation and network configuration for the Pod

> by the runtime plugin. Kubelet can start pulling container images and create

Reviewer 2: ```suggestion
by the runtime plugin. The kubelet can start pulling container images and create
```

containers after `PodHasNetwork` condition has been set to `True`.

> For a Pod with init containers, Kubelet sets the `Initialized` condition to

Reviewer 2: ```suggestion
For a Pod with init containers, the kubelet sets the `Initialized` condition to
```

`True` after the init containers have successfully completed (which happens
after successful sandbox creation and network configuration by the runtime

> plugin). For a Pod without init containers, Kubelet sets the `Initialized`

Reviewer 2: ```suggestion
plugin). For a Pod without init containers, the kubelet sets the `Initialized`
```
